From 6302c2ea52c6c28d36b6006b29823c469e171e2a Mon Sep 17 00:00:00 2001
From: Wen Liang <liangwen12year@gmail.com>
Date: Thu, 3 Aug 2023 10:16:42 -0400
Subject: [PATCH] nm-manager: ensure device is exported on D-Bus in
 authentication request

The device authentication request is an async process, it can not know
the answer right away, it is not guarantee that device is still
exported on D-Bus when authentication finishes. Thus, do not return
SUCCESS and abort the authentication request when device is not alive.

https://bugzilla.redhat.com/show_bug.cgi?id=2210271
(cherry picked from commit b341161e2a4988403ae4a6ef7653e01567da36a0)
(cherry picked from commit 0e27e84247ed824b27d105292d7bf42dc0341cbb)
---
 src/core/nm-manager.c | 15 +++++++++++++++
 1 file changed, 15 insertions(+)

diff --git a/src/core/nm-manager.c b/src/core/nm-manager.c
index 9c7212202b..937acbba1e 100644
--- a/src/core/nm-manager.c
+++ b/src/core/nm-manager.c
@@ -3222,6 +3222,13 @@ device_auth_done_cb(NMAuthChain *chain, GDBusMethodInvocation *context, gpointer
         nm_assert(error || (result == NM_AUTH_CALL_RESULT_YES));
     }
 
+    if (!error && !nm_dbus_object_is_exported(NM_DBUS_OBJECT(device))) {
+        g_set_error(&error,
+                    NM_MANAGER_ERROR,
+                    NM_MANAGER_ERROR_UNKNOWN_DEVICE,
+                    "device no longer exists");
+    }
+
     callback(device, context, subject, error, nm_auth_chain_get_data(chain, "user-data"));
 }
 
@@ -3287,6 +3294,14 @@ nm_manager_device_auth_request(NMManager                     *self,
                                                 &error))
         goto fail_on_idle;
 
+    if (!nm_dbus_object_is_exported(NM_DBUS_OBJECT(device))) {
+        g_set_error(&error,
+                    NM_MANAGER_ERROR,
+                    NM_MANAGER_ERROR_UNKNOWN_DEVICE,
+                    "device no longer exists");
+        goto fail_on_idle;
+    }
+
     chain = nm_auth_chain_new_subject(subject, context, device_auth_done_cb, self);
     if (cancellable)
         nm_auth_chain_set_cancellable(chain, cancellable);
-- 
2.41.0

